zkvm/src/vm.rs

```diff
@@ -331,9 +331,41 @@ where
         Ok(())
     }
 
+    fn unblind(&mut self) -> Result<(), VMError> {
+        // Pop expression `expr`
+        let expr = self.pop_item()?.to_expression()?;
+
+        // Pop commitment `V`
+        let v_commitment = self.pop_item()?.to_data()?.to_commitment()?;
+        let v_point = v_commitment.to_point();
+
+        // Pop scalar `v`
+        let scalar_witness = self.pop_item()?.to_data()?.to_scalar()?;
+        let v_scalar = scalar_witness.to_scalar();
+
+        self.delegate.verify_point_op(|| {
+            // Check V = vB => V-vB = 0
+            PointOp {
+                primary: Some(-v_scalar),
+                secondary: None,
+                arbitrary: vec![(Scalar::one(), v_point)],
+            }
+        })?;
+
+        // Add constraint `V == expr`
+        let (_, v) = self.delegate.commit_variable(&v_commitment)?;
+        self.delegate.cs().constrain(expr.to_r1cs_lc() - v);
+
+        // Push variable
+        self.push_item(Variable {
+            commitment: v_commitment,
+        });
+        Ok(())
+    }
+
     fn r#const(&mut self) -> Result<(), VMError> {
-        let data = self.pop_item()?.to_data()?.to_bytes();
-        let scalar = SliceReader::parse(&data, |r| r.read_scalar())?;
+        let scalar_witness = self.pop_item()?.to_data()?.to_scalar()?;
+        let scalar = scalar_witness.to_scalar();
```

Comment:
This line kills the integer witness. We need to pass ScalarWitness as is to Expression, to preserve its structure (integerness).

```suggestion
```

Reply:
ahh right, sorry missed that